virtual_rainforest/models/hydrology/below_ground.py

```diff
@@ -195,3 +199,77 @@ def convert_soil_moisture_to_water_potential(
     return air_entry_water_potential * (
         (soil_moisture / soil_moisture_capacity) ** water_retention_curvature
     )
+
+
+def update_groundwater_storge(
+    groundwater_storage: NDArray[np.float32],
+    vertical_flow_to_groundwater: NDArray[np.float32],
+    bypass_flow: NDArray[np.float32],
+    max_percolation_rate_uzlz: Union[float, NDArray[np.float32]],
+    groundwater_loss: Union[float, NDArray[np.float32]],
+    reservoir_const_upper_groundwater: Union[float, NDArray[np.float32]],
+    reservoir_const_lower_groundwater: Union[float, NDArray[np.float32]],
+) -> dict[str, NDArray[np.float32]]:
+    """Update groundwater storage and calculate below ground horizontal flow.
+
+    Groundwater storage and transport are modelled using two parallel linear reservoirs,
+    similar to the approach used in the HBV-96 model
+    :cite:p:`lindstrom_development_1997` and the LISFLOOD
+    :cite:p:`van_der_knijff_lisflood_2010`.
+    The upper zone represents a quick runoff component, which includes
+    fast groundwater and subsurface flow through macro-pores in the soil. The lower zone
+    represents the slow groundwater component that generates the base flow.
+
+    Args:
+        groundwater_storage: amount of water that is stored in the groundwater reservoir
+            , [mm]
+        vertical_flow_to_groundwater: flux from the lower soil layer to groundwater for
+            this timestep, [mm]
+        bypass_flow: flow that bypasses the soil matrix and drains directly to the
+            groundwater, [mm]
+        max_percolation_rate_uzlz: maximum perclation rate between pper and lower
+            groundwater zone, [mm d-1]
+        groundwater_loss: constant amount of water that never rejoins the river channel
+            and is lost beyond the catchment boundaries or to deep groundwater systems,
+            [mm]
+        reservoir_const_upper_groundwater: reservoir constant for the upper groundwater
+            layer, [days]
+        reservoir_const_lower_groundwater: reservoir constant for the lower groundwater
+            layer, [days]
+
+    Returns:
+        updated amount of water stored in upper and lower zone, outflow from the upper
+            zone to the channel, and outflow from the lower zone to the channel
+    """
+
+    output = {}
+    # percolation to lower zone
+    percolation_to_lower_zone = np.where(
+        max_percolation_rate_uzlz < groundwater_storage[0],
+        max_percolation_rate_uzlz,
+        groundwater_storage[0],
+    )
+
+    # Update water stored in upper zone, [mm]
+    upper_zone = np.array(
+        groundwater_storage[0]
+        + vertical_flow_to_groundwater
+        + bypass_flow
+        - percolation_to_lower_zone
+    )
+
+    # Calculate outflow from the upper zone to the channel, [mm]
+    output["subsurface_flow"] = 1 / reservoir_const_upper_groundwater * upper_zone
```

Comment:
I'd usually use brackets in cases like this for clarity, but in this case you can do:

```suggestion
    output["subsurface_flow"] = upper_zone / reservoir_const_upper_groundwater
```